### PR TITLE
promote for fma

### PR DESCRIPTION
**Pull request by jtlap**

### CODE REVIEW

File: include/eve/module/core/decorator/promote.hpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  SPDX-License-Identifier: BSL-1.0
+*/
+//==================================================================================================
+#pragma once
+
+#include <eve/detail/overload.hpp>
+
+namespace eve
+{
+template<auto Param> struct diff_;
```
Comment (jfalcou) [inline include/eve/module/core/decorator/promote.hpp:14]: Stop with the copy paste.

File: include/eve/module/core/decorator/promote.hpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+//================================================================================================
+//================================================================================================
+// Function decorators mark-up used in function overloads
+struct promote_
+{
+  template<auto N> static constexpr auto combine(decorated<diff_<N>()> const&) noexcept
```
Comment (jfalcou) [inline include/eve/module/core/decorator/promote.hpp:20]: diff is not even there anymore in EVE

File: include/eve/module/core/promote/impl/fma.hpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  {
+   using et_t =  element_type_t<T>;
+   using eu_t =  element_type_t<U>;
+   using ev_t =  element_type_t<V>;
+   using er_t = common_type_t<et_t, eu_t, ev_t>;
+   constexpr int ct = cardinal_v<T>;
+   constexpr int cu = cardinal_v<U>;
+   constexpr int cv = cardinal_v<V>;
+   constexpr int Ntu = ct < cu ? cu : cv;
+   constexpr int N   = cv < Ntu ? Ntu : cv;
```
Comment (jfalcou) [inline include/eve/module/core/promote/impl/fma.hpp:31]: why this comparaison ?

File: test/doc/core/regular/convert.cpp [hunk 4]
```diff
@@ -14,7 +14,7 @@ int main()
 
   std::cout << "---- simd" << '\n'
             << "<- pf                                 = " << pf << '\n'
-            << "-> convert(pf, int_64()_)             = " << eve::convert(pf, int_64()) << '\n'
+            << "-> convert(pf, int_64())              = " << eve::convert(pf, int_64()) << '\n'
```
Comment (jfalcou) [inline test/doc/core/regular/convert.cpp:17]: is this file even tested ?

File: test/unit/module/core/fma.cpp [hunk 5]
```diff
@@ -38,7 +38,7 @@ TTS_CASE_TPL("Check return types of fma", eve::test::simd::all_types)
 };
 
 //==================================================================================================
-// fma tests
+//== fma tests
```
Comment (jfalcou) [inline test/unit/module/core/fma.cpp:41]: no

File: test/unit/module/core/fma.cpp [hunk 6]
```diff
@@ -63,49 +63,61 @@ TTS_CASE_WITH("Check precision behavior of fma on real types",
...
+  eve::wide<int, eve::fixed<N>> ia([](auto i,  auto){return int(i); });
+  auto r3 = promote(fma)(a0, a1, ia);
+  using er3_t =  eve::element_type_t<decltype(r3)>;
+  auto refr3 = eve::fma(eve::convert(a0, eve::as<er3_t>()), eve::convert(a1, eve::as<er3_t>()), eve::convert(ia, eve::as<er3_t>()));
+  TTS_EQUAL(r3,  refr3);
 };
 
 //==================================================================================================
-// fma masked
+//== fma masked
```
Comment (jfalcou) [inline test/unit/module/core/fma.cpp:120]: why this noise ?

**Review by jfalcou — CHANGES_REQUESTED** [review]

File: test/unit/module/core/fma.cpp (hunk 6 above)
Comment (jtlap) [reply test/unit/module/core/fma.cpp:120]: commenting to test faster then uncommenting wa not idempotent

File: include/eve/module/core/decorator/promote.hpp (hunk 1 above)
Comment (jtlap) [reply include/eve/module/core/decorator/promote.hpp:14]: I thought that pedantic was ok
I had not seen diff

File: include/eve/module/core/decorator/promote.hpp (hunk 2 above)
Comment (jfalcou) [reply include/eve/module/core/decorator/promote.hpp:20]: this is still not needed

Comment (jtlap) [reply include/eve/module/core/decorator/promote.hpp:20]: i just copied pedantic et là j'ai enlevé la ligne

File: include/eve/module/core/promote/impl/fma.hpp (hunk 3 above)
Comment (jtlap) [reply include/eve/module/core/promote/impl/fma.hpp:31]: parce que certains des  types peuvent etre scalaires et d'autres simd

File: include/eve/module/core/decorator/promote.hpp (hunk 2 above)
Comment (jfalcou) [reply include/eve/module/core/decorator/promote.hpp]: You removed the whole decorator, that is not the issue.

File: include/eve/module/core/promote/impl/fma.hpp (hunk 3 above)
Comment (jfalcou) [inline include/eve/module/core/promote/impl/fma.hpp:31]: `constexpr auto N   = std::max({cardinal_v<T>,cardinal_v<U>,cardinal_v<V>});`
looks better.

Need to include `<algorithm>`.

**Review by jfalcou — CHANGES_REQUESTED** [review]

File: test/doc/core/regular/convert.cpp (hunk 4 above)
Comment (jfalcou) [reply test/doc/core/regular/convert.cpp:17]: Oh OK that's just the display

File: include/eve/module/core/promote/impl/fma.hpp (hunk 3 above)
Comment (jtlap) [reply include/eve/module/core/promote/impl/fma.hpp]: done

**Review by jfalcou — APPROVED** [review]

**Pull request merged**